src/fixate/drivers/daq/daqmx.py

```diff
@@ -392,31 +396,44 @@ def init(self):
             self.task_state = "init"
 
     def read(self):
-        self._trigger_thread.join(10)
-        if self._trigger_thread.exec_info:
-            try:
-                raise self._trigger_thread.exec_info
-            finally:
-                self._trigger_thread = None
-        self._trigger_thread = None
+        self._trigger_thread.join(self._thread_timeout)
+        if self._trigger_thread.is_alive():
+            raise InstrumentError("Trigger thread failed to terminate")
+        try:
+            err = self._error_queue.get_nowait()
+        except Empty:
+            # no error in queue
+            pass
+        else:
+            raise err
         return self._data
 
     def _read(self):
-        self.init()
-        return DAQmxReadCounterScalarF64(
-            self.task, float64(10), byref(self._data), None
-        )
+        try:
+            self.init()
+            DAQmxReadCounterScalarF64(self.task, float64(10), byref(self._data), None)
+        except Exception as e:
+            self._error_queue.put(ThreadError(e))
+        return
 
     def trigger(self):
         if self._trigger_thread:
             self.clear()
-            self._trigger_thread.join(10)
+            self._trigger_thread.join(self._thread_timeout)
```

Comment:
I really wish NI had better documentation... It not at all clear how to correctly use the Task api. It feels to me like the clear() and join() calls here are unnecessary. I think the clear(), which call DAQmxClearTask() actually deallocates the task. So this method should never work if the self._trigger_thread block executes...

But I am reluctant to dive too far into this right now. My gut feel is that a nice way to use the daq tasks would be to implement them as a context manager, so that clean up is done correctly.

Reply:
Clearing the task does deallocate the task, and sets the state to "", which means that when the trigger thread is created the init method recreates the task again.
I'm not sure whether the init should be done in the thread itself or whether it should be done in the trigger call.
I think that reading the result is really the only thing that needs to be in the thread. I can test it to be sure if needed.